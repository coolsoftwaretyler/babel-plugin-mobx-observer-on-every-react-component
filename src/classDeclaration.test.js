import { expect, test, describe } from "bun:test";
import { transform } from "@babel/core";
import pluginSyntaxJsx from "@babel/plugin-syntax-jsx";
import pluginTransReactJsx from "@babel/plugin-transform-react-jsx";
import pluginSyntaxDecorators from "@babel/plugin-syntax-decorators";
import pluginSyntaxDecoratorsLegacy from "@babel/plugin-proposal-decorators";
import autoObserverPlugin from "./index.ts";

const runTransform = (input) => {
  return transform(input, {
    plugins: [pluginSyntaxJsx, pluginTransReactJsx, [pluginSyntaxDecorators, { version: "2018-09", decoratorsBeforeExport: false }], autoObserverPlugin],
  });
};

const runTransformLegacy = (input) => {
  return transform(input, {
    plugins: [pluginSyntaxJsx, pluginTransReactJsx, [pluginSyntaxDecoratorsLegacy, { legacy: true }], autoObserverPlugin],
  });
};

describe("Mobx Observer Babel Plugin", () => {
  describe("class declarations", () => {
    describe("when it is not a React component", () => {
      test("it does nothing", () => {
        const source = `class MyComponentClass {}`;
        const out = runTransform(source);

        expect(out.code).toMatchSnapshot();
      });
    });
    describe("when it is a React component", () => {
      describe("and it is already wrapped in observer", () => {
        test("it does nothing", () => {
          const source = `observer(class MyComponent { render() { return <div>Hello World</div>; } })`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe('and it is decorated with @observer', () => {
        describe('modern', () => {
          test('it does nothing', () => {
            const source = `@observer class MyComponent { render() { return <div>Hello World</div>; } }`;
            const out = runTransform(source);

            console.log(out.code);

            expect(out.code).toMatchSnapshot();
          });
        })
        
        describe('legacy', () => {
          test('it does nothing', () => {
            const source = `@observer class MyComponent { render() { return <div>Hello World</div>; } }`;
            const out = runTransformLegacy(source);

            expect(out.code).toMatchSnapshot();
          });
        });
      });
      describe("and it is not wrapped in observer", () => {
        test("it wraps the component in observer", () => {
          const source = `class MyComponent { render() { return <div>Hello World</div>; } }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
    });
  });
});
