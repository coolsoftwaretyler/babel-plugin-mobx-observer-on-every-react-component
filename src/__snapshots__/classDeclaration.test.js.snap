// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobx Observer Babel Plugin class declarations when it is not a React component it does nothing 1`] = `"class MyComponentClass {}"`;

exports[`Mobx Observer Babel Plugin class declarations when it is a React component and it is already wrapped in observer it does nothing 1`] = `
"observer(class MyComponent {
  render() {
    return /*#__PURE__*/React.createElement("div", null, "Hello World");
  }
});"
`;

exports[`Mobx Observer Babel Plugin class declarations when it is a React component and it is not wrapped in observer it wraps the component in observer 1`] = `
"observer(class MyComponent {
  render() {
    return /*#__PURE__*/React.createElement("div", null, "Hello World");
  }
});"
`;
